File: app/src/main/java/com/tunjid/fingergestures/activities/PopupActivity.java
package com.tunjid.fingergestures.activities;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;
import android.view.Window;
import android.widget.TextView;

import com.google.android.material.card.MaterialCardView;
import com.tunjid.androidbootstrap.recyclerview.ListManager;
import com.tunjid.androidbootstrap.recyclerview.ListManagerBuilder;
import com.tunjid.fingergestures.App;
import com.tunjid.fingergestures.BackgroundManager;
import com.tunjid.fingergestures.PopUpGestureConsumer;
import com.tunjid.fingergestures.R;
import com.tunjid.fingergestures.adapters.ActionAdapter;
import com.tunjid.fingergestures.gestureconsumers.GestureConsumer;
import com.tunjid.fingergestures.gestureconsumers.GestureMapper;
import com.tunjid.fingergestures.viewholders.ActionViewHolder;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.GridLayoutManager.SpanSizeLookup;
import io.reactivex.disposables.CompositeDisposable;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;
import static android.view.ViewGroup.LayoutParams.MATCH_PARENT;

public class PopupActivity extends AppCompatActivity {

    private CompositeDisposable disposables;

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        finish();
    }

    @SuppressLint("ClickableViewAccessibility")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_popup);

        disposables = new CompositeDisposable();

        Window window = getWindow();
        window.setLayout(MATCH_PARENT, MATCH_PARENT);

        List<Integer> items = new ArrayList<>();
        AtomicInteger spanSizer = new AtomicInteger(0);
        BackgroundManager backgroundManager = BackgroundManager.getInstance();

        ListManager<ActionViewHolder, Void> listManager = new ListManagerBuilder<ActionViewHolder, Void>()
                .withRecyclerView(findViewById(R.id.item_list))
                .withGridLayoutManager(6)
                .withAdapter(new ActionAdapter(true, true, items, this::onActionClicked))
                .onLayoutManager(manager -> ((GridLayoutManager) manager).setSpanSizeLookup(new SpanSizeLookup() {
                    @Override public int getSpanSize(int position) { return spanSizer.get(); }
                }))
                .build();

        TextView text = findViewById(R.id.text);
        text.setTextColor(backgroundManager.getSliderColor());
        this.<MaterialCardView>findViewById(R.id.card).setCardBackgroundColor(backgroundManager.getBackgroundColor());

        disposables.add(App.diff(items, PopUpGestureConsumer.getInstance()::getList).subscribe(result -> {
            int size = items.size();
            listManager.onDiff(result);
            spanSizer.set(size == 1 ? 6 : size == 2 ? 3 : 2);
            text.setVisibility(size == 0 ? VISIBLE : GONE);
        }, Throwable::printStackTrace));

        findViewById(R.id.constraint_layout).setOnTouchListener((view, motionEvent) -> {
            finish();
            return true;
        });
    }

    protected void onResume() {
        super.onResume();

        if (PopUpGestureConsumer.getInstance().shouldAnimatePopup())
            overridePendingTransition(R.anim.slide_in_up, R.anim.slide_out_down);
    }

    @Override
    protected void onStop() {
        super.onStop();
        finish();
    }

    @Override
    public void finish() {
        super.finish();
        boolean shouldAnimate = PopUpGestureConsumer.getInstance().shouldAnimatePopup();
        if (shouldAnimate) overridePendingTransition(R.anim.slide_in_up, R.anim.slide_out_down);
        else overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out);
        disposables.clear();
    }

    private void onActionClicked(@GestureConsumer.GestureAction int action) {
        finish();
        GestureMapper.getInstance().performAction(action);
    }
}

File: app/src/main/java/com/tunjid/fingergestures/activities/PopupActivity.kt
/*
 * Copyright (c) 2017, 2018, 2019 Adetunji Dahunsi.
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.tunjid.fingergestures.activities

import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import android.view.ViewGroup.LayoutParams.MATCH_PARENT
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.isVisible
import com.tunjid.androidx.recyclerview.gridLayoutManager
import com.tunjid.androidx.recyclerview.listAdapterOf
import com.tunjid.androidx.view.util.inflate
import com.tunjid.fingergestures.BackgroundManager
import com.tunjid.fingergestures.PopUpGestureConsumer
import com.tunjid.fingergestures.R
import com.tunjid.fingergestures.databinding.ActivityPopupBinding
import com.tunjid.fingergestures.map
import com.tunjid.fingergestures.models.Action
import com.tunjid.fingergestures.viewholders.ActionViewHolder
import com.tunjid.fingergestures.viewmodels.*
import java.util.concurrent.atomic.AtomicInteger

class PopupActivity : AppCompatActivity() {

    private val binding by lazy { ActivityPopupBinding.inflate(layoutInflater) }
    private val viewModel by viewModels<PopUpViewModel>()

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        finish()
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(binding.root)

        val window = window
        window.setLayout(MATCH_PARENT, MATCH_PARENT)

        val spanSizer = AtomicInteger(0)
        val backgroundManager = BackgroundManager.instance

        val listAdapter = listAdapterOf(
            initialItems = listOf<Action>(),
            viewHolderCreator = { viewGroup, _ ->
                ActionViewHolder(
                    showsText = true,
                    itemView = viewGroup.inflate(R.layout.viewholder_action_horizontal),
                    clickListener = ::onActionClicked
                )
            },
            viewHolderBinder = { holder, item, _ -> holder.bind(item) }
        )

        binding.text.setTextColor(backgroundManager.sliderColorPreference.value)
        binding.card.setCardBackgroundColor(backgroundManager.backgroundColorPreference.value)
        binding.itemList.apply {
            layoutManager = gridLayoutManager(6) { spanSizer.get() }
            adapter = listAdapter
        }
        binding.constraintLayout.setOnTouchListener { _, _ ->
            finish()
            true
        }

        viewModel.state.map(PopUpState::popUpActions).observe(this@PopupActivity) { items ->
            val size = items.size
            spanSizer.set(if (size == 1) 6 else if (size == 2) 3 else 2)
            binding.text.isVisible = size == 0
            listAdapter.submitList(items)
        }
    }

    override fun onResume() {
        super.onResume()

        if (PopUpGestureConsumer.instance.animatePopUpPreference.value)
            overridePendingTransition(R.anim.slide_in_up, R.anim.slide_out_down)
    }

    override fun onStop() {
        super.onStop()
        finish()
    }

    override fun finish() {
        super.finish()
        val shouldAnimate = PopUpGestureConsumer.instance.animatePopUpPreference.value
        if (shouldAnimate) overridePendingTransition(R.anim.slide_in_up, R.anim.slide_out_down)
        else overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out)
    }

    private fun onActionClicked(action: Action) {
        viewModel.accept(PopUpInput.Perform(action))
        finish()
    }
}
